sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -1235,11 +1235,13 @@ object TryToTimestampExpressionBuilder extends ExpressionBuilder {
   override def build(funcName: String, expressions: Seq[Expression]): Expression = {
     val numArgs = expressions.length
     if (numArgs == 1 || numArgs == 2) {
-      ParseToTimestamp(
+      // The expression ParseToTimestamp will throw an SparkUpgradeException if the input is invalid
+      // even when failOnError is false. We need to catch the exception and return null.
+      TryEval(ParseToTimestamp(
```

Comment:
shall we simply update `ToTimestamp#eval` and codegen to catch `SparkUpgradeException` additionally? I'm a bit worried about increasing the scope and swallow all errors.

Reply:
So in `ToTimestamp`, there is supposed to be a `SparkUpgradeException` even when `failOnError`is false.
To avoid swallowing all errors, we can add another flag `isTryFunction` besides `failOnError`. It will make the code more complicated though.

Reply:
so we need to maintain 3 behaviors:
1. `try_to_timestamp` never fails
2. `to_timestamp` ansi mode always fail with invalid input
3. `to_timestamp` non-ansi mode only fails with `SparkUpgradeException`

Do I understand it correctly?

Reply:
I'm not sure why `try_to_timestamp` need a different behavior with `to_timestamp` non-ansi mode.

Reply:
> I'm not sure why try_to_timestamp need a different behavior with to_timestamp non-ansi mode.

The try_* functions are supposed to return null on invalid inputs, instead of compatible with non-ansi mode. cc @srielau on this one.

Reply:
This upgrade issue seems special and we should always throw it to provide upgrade instructions. How about https://github.com/apache/spark/pull/45853 ?

Reply:
@cloud-fan after https://github.com/apache/spark/pull/45853, users will face this bug when setting the timeParserPolicy as legacy.

Reply:
We can argue this is not a bug as the upgrade exception must be thrown.

Reply:
@cloud-fan ok, let me close this one for now.